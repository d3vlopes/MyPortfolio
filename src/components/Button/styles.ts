import styled, { css, DefaultTheme } from 'styled-components'

import { ButtonProps } from '.'

export type WrapperProps = {
  expand: boolean
  hasIcon: boolean
} & ButtonProps

const wrapperModifiers = {
  primary: (theme: DefaultTheme) => css`
    background-color: ${theme.colors.primary.main};

    &:hover {
      background-color: ${theme.colors.primary.dark};
    }
  `,
  secondary: (theme: DefaultTheme) => css`
    background-color: ${theme.colors.secondary.main};

    &:hover {
      background-color: ${theme.colors.secondary.dark};
    }
  `,
  outline: (theme: DefaultTheme) => css`
    background-color: transparent;
    border: 1px solid ${theme.colors.complementary.main};

    &:hover {
      span {
        color: ${theme.colors.primary.main};
        transition: all ${theme.transition.default};
      }

      border: 1px solid ${theme.colors.primary.main};
    }
  `,
  small: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.small};
    width: 17.5rem;
    height: 5rem;

    svg {
      width: 2.2rem;
      height: 2.2rem;
    }
  `,
  medium: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.large};
    width: 21.5rem;
    height: 6.2rem;

    svg {
      width: 2.4rem;
      height: 2.4rem;
    }
  `,
  large: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.xxlarge};
    width: 28.8rem;
    height: 7.6rem;

    svg {
      width: 2.8rem;
      height: 2.8rem;
    }
  `,
  expand: () => css`
    width: 31.5rem;
    height: 6.2rem;
  `,
  withIcon: (theme: DefaultTheme) => css`
    border-radius: ${theme.border.radius.large};
    height: 5rem;

    svg {
      color: ${theme.colors.complementary.main};
      transition: all ${theme.transition.default};
    }
  `,
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, variant, size, expand, hasIcon, hoverUp }) => css`
    display: inline-flex;
    justify-content: center;
    align-items: center;
    border: none;
    transition: all ${theme.transition.default};
    text-decoration: none;
    cursor: pointer;
    border-radius: ${theme.border.radius.medium};

    &:focus {
      outline: 1px dashed;
    }

    &:hover {
      transform: ${hoverUp ? 'translateY(-0.5rem)' : 'translateY(0)'};

      svg {
        color: ${theme.colors.primary.main};
      }
    }

    span {
      font-family: ${theme.font.body};
      font-weight: ${theme.font.weight.medium};
      line-height: 100%;
      color: ${theme.colors.text.light};
      margin-left: ${hasIcon && theme.spacings.xxsmall};
      letter-spacing: 0.05rem;
    }

    ${!!variant && wrapperModifiers[variant](theme)};
    ${!!size && wrapperModifiers[size](theme)};
    ${expand && wrapperModifiers.expand()};
    ${hasIcon && wrapperModifiers.withIcon(theme)};
  `}
`
