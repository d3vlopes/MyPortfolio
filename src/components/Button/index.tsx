import { forwardRef, AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react'

import * as S from './styles'

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  variant?: 'primary' | 'secondary' | 'outline'
  size?: 'small' | 'medium' | 'large'
  expand?: boolean
  icon?: JSX.Element
  as?: React.ElementType
} & ButtonTypes

const Button: React.ForwardRefRenderFunction<S.WrapperProps, ButtonProps> = (
  {
    children,
    icon,
    size = 'medium',
    expand = false,
    variant = 'primary',
    ...props
  },
  ref,
) => (
  <S.Wrapper
    size={size}
    expand={expand}
    hasIcon={!!icon}
    variant={variant}
    ref={ref}
    {...props}
  >
    {icon}

    {!!children && <span>{children}</span>}
  </S.Wrapper>
)

export default forwardRef(Button)
