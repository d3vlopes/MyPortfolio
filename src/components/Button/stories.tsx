import { Story, Meta } from '@storybook/react'
import { Github } from '@styled-icons/boxicons-logos'

import Button, { ButtonProps } from '.'

export default {
  title: 'Components/Button',
  component: Button,
  parameters: {
    layout: 'centered',
  },
  args: {
    size: 'medium',
    expand: false,
    hoverUp: false,
  },
  argTypes: {
    icon: {
      type: '',
    },
    variant: {
      control: {
        type: 'select',
        options: ['primary', 'secondary', 'outline'],
      },
    },
    size: {
      control: {
        type: 'select',
        options: ['small', 'medium', 'large'],
      },
    },
  },
} as Meta<ButtonProps>

export const Primary: Story<ButtonProps> = (args) => {
  return (
    <div>
      <Button {...args} />
    </div>
  )
}

Primary.args = {
  children: 'Button',
  variant: 'primary',
}

export const Secondary: Story<ButtonProps> = (args) => {
  return (
    <div>
      <Button {...args} />
    </div>
  )
}

Secondary.args = {
  children: 'Button',
  variant: 'secondary',
  expand: false,
}

export const Expand: Story<ButtonProps> = (args) => {
  return (
    <div>
      <Button {...args} />
    </div>
  )
}

Expand.args = {
  children: 'Button',
  variant: 'primary',
  expand: true,
}

export const withIcon: Story<ButtonProps> = (args) => {
  return (
    <div>
      <Button {...args} />
    </div>
  )
}

withIcon.args = {
  children: 'Ver c√≥digo',
  variant: 'outline',
  icon: <Github />,
}
