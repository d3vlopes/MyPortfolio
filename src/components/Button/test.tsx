import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test/helpers'
import { Github } from '@styled-icons/boxicons-logos'

import Button from '.'

const { getByRole, getByTestId } = screen

describe('<Button />', () => {
  it('should render by default', () => {
    renderWithTheme(
      <Button variant="primary" size="medium">
        Button
      </Button>,
    )

    const button = getByRole('button', { name: /button/i })

    expect(button).toBeInTheDocument()
    expect(button).toHaveStyle({
      'background-color': '#6D59A8',
      'font-size': '2.2rem',
      width: '21.5rem',
      height: '6.2rem',
    })
  })

  it('should render the variant secondary ', () => {
    renderWithTheme(<Button variant="secondary">Button</Button>)

    const button = getByRole('button', { name: /button/i })

    expect(button).toBeInTheDocument()
    expect(button).toHaveStyle({
      'background-color': '#83A848',
    })
  })

  it('should render the variant outline ', () => {
    renderWithTheme(<Button variant="outline">Button</Button>)

    const button = getByRole('button', { name: /button/i })

    expect(button).toBeInTheDocument()
    expect(button).toHaveStyle({
      border: '1px solid #DEE7EA',
      'background-color': 'transparent',
    })
  })

  it('should render expand version', () => {
    renderWithTheme(<Button expand>Button</Button>)

    const button = getByRole('button', { name: /button/i })

    expect(button).toHaveStyle({
      width: '31.5rem',
      height: '6.2rem',
    })
  })

  it('should render small size', () => {
    renderWithTheme(<Button size="small">Button</Button>)

    const button = getByRole('button', { name: /button/i })

    expect(button).toHaveStyle({
      'font-size': '1.8rem',
      width: '17.5rem',
      height: '5rem',
    })
  })

  it('should render large size', () => {
    renderWithTheme(<Button size="large">Button</Button>)

    const button = getByRole('button', { name: /button/i })

    expect(button).toHaveStyle({
      'font-size': '2.6rem',
      width: '28.8rem',
      height: '7.6rem',
    })
  })

  it('should render an icon version', () => {
    renderWithTheme(
      <Button icon={<Github data-testid="icon" />}>Button</Button>,
    )

    const button = getByRole('button', { name: /button/i })
    const icon = getByTestId('icon')

    expect(button).toBeInTheDocument()
    expect(icon).toBeInTheDocument()
  })

  it('should render Button as a link', () => {
    renderWithTheme(
      <Button as="a" href="/link">
        Button
      </Button>,
    )

    expect(screen.getByRole('link', { name: /button/i })).toHaveAttribute(
      'href',
      '/link',
    )
  })

  it('should up while hover', async () => {
    renderWithTheme(<Button hoverUp>Button</Button>)

    const button = getByRole('button', { name: /button/i })

    expect(button).toHaveStyleRule('transform', 'translateY(-0.5rem)', {
      modifier: ':hover',
    })
  })

  it('should match snapshot', () => {
    const { container } = renderWithTheme(<Button>Button</Button>)

    expect(container.firstChild).toMatchSnapshot()
  })
})
