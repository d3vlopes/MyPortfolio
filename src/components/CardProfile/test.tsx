import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test/helpers'

import CardProfile, { CardProfileProps } from '.'

const props: CardProfileProps = {
  image: '/img/me.jpg',
  name: 'Leandro Lopes',
  age: 27,
  city: 'Santa Cruz do Sul - RS',
  liking: 'Programação, design, música, séries e filmes',
  dream:
    'Trabalhar com o que amo, utilizando meu conhecimento para criar aplicações que possam ajudar as pessoas.',
}

const { getByRole, getByText } = screen

describe('<CardProfile />', () => {
  it('should render CardProfile', () => {
    const { container } = renderWithTheme(<CardProfile {...props} />)

    const image = getByRole('img', { name: 'Leandro Lopes' })
    const name = getByText('Leandro Lopes')
    const age = getByText('27 anos')
    const city = getByText('Santa Cruz do Sul - RS')
    const liking = getByText('Programação, design, música, séries e filmes')
    const dream = getByText(
      'Trabalhar com o que amo, utilizando meu conhecimento para criar aplicações que possam ajudar as pessoas.',
    )

    expect(image).toHaveAttribute('src', '/img/me.jpg')
    expect(name).toBeInTheDocument()
    expect(age).toBeInTheDocument()
    expect(city).toBeInTheDocument()
    expect(liking).toBeInTheDocument()
    expect(dream).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
