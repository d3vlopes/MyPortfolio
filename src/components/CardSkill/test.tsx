import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test/helpers'
import { Sword } from '@styled-icons/remix-fill/Sword'

import CardSkill from '.'
import mock from './mock'

const { getByTestId, getByRole, getByText } = screen

describe('<CardSkill />', () => {
  it('should render CardSkill', () => {
    renderWithTheme(
      <CardSkill
        icon={<Sword data-testid="icon" />}
        title={mock[0].title}
        description={mock[0].description}
      />,
    )

    const icon = getByTestId('icon')
    const title = getByRole('heading', { name: /desafios/i })
    const description = getByText(
      /Estou sempre em busca de novos desafios que me tirem do meu porto seguro./i,
    )

    expect(icon).toBeInTheDocument()
    expect(title).toBeInTheDocument()
    expect(description).toBeInTheDocument()
  })

  it('should match snapshot', () => {
    const { container } = renderWithTheme(
      <CardSkill
        icon={<Sword data-testid="icon" />}
        title={mock[0].title}
        description={mock[0].description}
      />,
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
