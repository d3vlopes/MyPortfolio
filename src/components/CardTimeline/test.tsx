import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test/helpers'

import CardTimeline, { CardTimelineProps } from '.'

const props: CardTimelineProps = {
  title: 'CardTimeline',
  content: '<p>Loren ipsum dolor</p>',
}

const { getByRole, getByText } = screen

describe('<CardTimeline />', () => {
  it('should render CardTimeline', () => {
    const { container } = renderWithTheme(<CardTimeline {...props} />)

    const title = getByRole('heading', { name: 'CardTimeline' })
    const content = getByText('Loren ipsum dolor')
    const p = container.querySelector('p')

    expect(title).toBeInTheDocument()
    expect(content).toBeInTheDocument()
    expect(p?.tagName.toLowerCase()).toBe('p')
  })

  it('should match snapshot ', () => {
    const { container } = renderWithTheme(<CardTimeline {...props} />)

    expect(container.firstChild).toMatchSnapshot()
  })
})
