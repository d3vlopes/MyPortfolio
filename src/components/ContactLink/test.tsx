import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test/helpers'
import { Props } from 'react-inlinesvg'

import ContactLink, { ContactLinkProps } from '.'

const { getByLabelText } = screen

const mock: ContactLinkProps = {
  icon: '/img/icon.svg',
  name: 'Github',
  link: 'www.github.com/user/name',
}

jest.mock('react-inlinesvg', () => {
  function ReactInlineSVG({ name }: Props) {
    return <svg aria-label={name} />
  }

  return ReactInlineSVG
})

describe('<ContactLinks />', () => {
  it('should render ContactLinks', () => {
    const { container } = renderWithTheme(<ContactLink {...mock} />)

    const icon = getByLabelText(mock.name)
    const a = container.querySelector('a')

    expect(icon).toBeInTheDocument()
    expect(a).toHaveAttribute('href', mock.link)
  })
})
