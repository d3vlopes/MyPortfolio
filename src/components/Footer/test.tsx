import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test/helpers'

import mock from 'components/ContactLink/mock'

import Footer, { FooterProps } from '.'
import { Props } from 'react-inlinesvg/esm'

const props: FooterProps = {
  contactLinks: mock,
}

const { getByLabelText, getByTestId } = screen

jest.mock('react-inlinesvg', () => {
  function ReactInlineSVG({ name }: Props) {
    return <svg aria-label={name} />
  }

  return ReactInlineSVG
})

describe('<Footer />', () => {
  it('should render Footer', () => {
    const { container } = renderWithTheme(<Footer {...props} />)

    const contactLinks = getByTestId('contact-links')
    const text = container.querySelector('p')
    const heartIcon = getByLabelText(/heart/i)

    expect(contactLinks).toBeInTheDocument()
    expect(text).toHaveTextContent(/Desenvolvido com por Leandro Lopes/i)
    expect(heartIcon).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
