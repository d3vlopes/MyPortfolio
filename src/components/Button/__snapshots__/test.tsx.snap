// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should match snapshot 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  -webkit-transition: background 0.1s ease-in-out,-webkit-transform 0.3s ease-in-out;
  -webkit-transition: background 0.1s ease-in-out,transform 0.3s ease-in-out;
  transition: background 0.1s ease-in-out,transform 0.3s ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  border-radius: 0.6rem;
  background-color: #6D59A8;
  font-size: 2.2rem;
  width: 21.5rem;
  height: 6.2rem;
}

.c0:focus {
  outline: 1px dashed #DEE7EA;
}

.c0:hover {
  -webkit-transform: translateY(0);
  -ms-transform: translateY(0);
  transform: translateY(0);
}

.c0:hover svg {
  color: #6D59A8;
}

.c0 span {
  font-family: Rubik,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-weight: 500;
  line-height: 100%;
  color: #FFFFFF;
  -webkit-letter-spacing: 0.05rem;
  -moz-letter-spacing: 0.05rem;
  -ms-letter-spacing: 0.05rem;
  letter-spacing: 0.05rem;
}

.c0:hover {
  background-color: #543F8C;
}

.c0 svg {
  width: 2.4rem;
  height: 2.4rem;
}

<button
  class="c0"
>
  <span>
    Button
  </span>
</button>
`
